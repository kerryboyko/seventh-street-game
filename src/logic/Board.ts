import { invert } from "radash";
import Deck from "./poker/Deck";
import Card from "./poker/Card";

// The parks will be at index 0,
// 1 - 7 will be the corresponding streets.

interface Property {
  name: string;
  card: null | Card;
  revealed: boolean;
  owner: string | null;
  scored: boolean;
  price: number;
  type: "ACTION" | "PROPERTY";
  visibleToThesePlayers: Set<string>;
}

interface OutputProperty
  extends Pick<Property, "name" | "revealed" | "owner" | "scored" | "price"> {
  cardDisplay: string;
}

const AVENUES = ["N", "E", "S", "W"];
const generateNeighborhood = (deck: Deck): Record<string, Property> => {
  deck.shuffle();
  const neighborhood: Record<string, Property> = {};
  for (let avenue of AVENUES) {
    for (let i = 0; i < 8; i++) {
      const name = `${avenue}${i}`;
      neighborhood[name] = {
        name,
        card: i === 0 ? null : deck.dealCard(),
        revealed: false,
        visibleToThesePlayers: new Set<string>([]),
        owner: null,
        scored: false,
        price: i * 100,
        type: i === 0 ? "ACTION" : "PROPERTY",
      };
    }
  }
  return neighborhood;
};
export default class Board {
  private deck: Deck;
  private neighborhood;
  public players: string[] = [];
  constructor(public seed?: string) {
    this.deck = new Deck(this.seed);
    this.neighborhood = generateNeighborhood(this.deck);
  }
  public debug = () => ({
    deck: this.deck,
    neighborhood: this.neighborhood,
  });
  public addPlayer = (playerName: string) => {
    if (this.players.includes(playerName)) {
      return false;
    }
    this.players.push(playerName);
    return true;
  };
  public getBoardState = (player?: string): Record<string, OutputProperty> => {
    const output: Record<string, OutputProperty> = {};
    for (const name in this.neighborhood) {
      let {
        card,
        revealed,
        visibleToThesePlayers,
        owner,
        scored,
        price,
        type,
      } = this.neighborhood[name];
      if (type === "ACTION" || card === null) {
        continue;
      }
      const visible = revealed || (player && visibleToThesePlayers.has(player));
      output[name] = {
        name,
        cardDisplay: visible ? card?.display : "??",
        price,
        scored,
        owner,
        revealed,
      };
    }
    return output;
  };
  public setOwner = (
    owner: string,
    propertyName: keyof Board["neighborhood"]
  ) => {
    if (!this.players.includes(owner)) {
      throw new Error(`No player named ${owner} is playing`);
    }
    if (!this.neighborhood[propertyName]) {
      throw new Error(`No property named ${propertyName} exists`);
    }
    this.neighborhood[propertyName].owner = owner;
    this.neighborhood[propertyName].visibleToThesePlayers.add(owner);
  };
}
