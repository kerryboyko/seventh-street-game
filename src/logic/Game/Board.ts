import Deck from "../poker/Deck";
import Card from "../poker/Card";
import Property from "./Property";
// The parks will be at index 0,
// 1 - 7 will be the corresponding streets.

const AVENUES = ["N", "E", "S", "W"];
const generateNeighborhood = (deck: Deck): Record<string, Property> => {
  deck.shuffle();
  const neighborhood: Record<string, Property> = {};
  for (let avenue of AVENUES) {
    for (let i = 0; i < 8; i++) {
      const name = `${avenue}${i}`;
      neighborhood[name] = new Property(
        name,
        i === 0 ? null : deck.dealCard(),
        i * 100,
        i === 0 ? "ACTION" : "PROPERTY"
      );
    }
  }
  return neighborhood;
};
export default class Board {
  private deck: Deck;
  private neighborhood: Record<string, Property>;
  public players: string[] = [];
  constructor(public seed?: string) {
    this.deck = new Deck(this.seed);
    this.neighborhood = generateNeighborhood(this.deck);
  }
  public debug = () => ({
    deck: this.deck,
    neighborhood: this.neighborhood,
  });
  public addPlayer = (playerName: string) => {
    if (this.players.includes(playerName)) {
      return false;
    }
    this.players.push(playerName);
    return true;
  };
  public getBoardState = (player?: string): Record<string, any> => {
    const acc: Record<string, any> = {};
    Object.values(this.neighborhood)
      .filter(({ type }) => type === "PROPERTY")
      .forEach((prop: Property) => {
        acc[prop.name] = prop.getState(player);
      });
      return acc;
  };

  public setOwner = (
    owner: string,
    propertyName: keyof Board["neighborhood"]
  ) => {
    if (!this.players.includes(owner)) {
      throw new Error(`No player named ${owner} is playing`);
    }
    if (!this.neighborhood[propertyName]) {
      throw new Error(`No property named ${propertyName} exists`);
    }
    this.neighborhood[propertyName].setOwner(owner);
  };
}
