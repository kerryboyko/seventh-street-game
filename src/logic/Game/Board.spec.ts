import Board from "./Board";
import { pick, mapValues } from "radash";

describe("class Board", () => {
  const board = new Board("seventhst test");
  it("constructor() generates a neighborhood", () => {
    expect(board.debug()).toMatchSnapshot();
  });
  it("Board.getBoardState() can get a board state", () => {
    expect(board.getBoardState()).toMatchSnapshot();
  });
  it("Board.addPlayer() can add players", () => {
    expect(board.addPlayer("player1")).toBe(true);
    expect(board.addPlayer("player2")).toBe(true);
    expect(board.players).toEqual(["player1", "player2"]);
    expect(board.addPlayer("player1")).toBe(false);
    expect(board.players).toEqual(["player1", "player2"]);
    expect(board.addPlayer("player3")).toBe(true);
    expect(board.players).toEqual(["player1", "player2", "player3"]);
  });
  it("Board.setOwner can set owners of properties", () => {
    board.setOwner("player1", "E1");
    board.setOwner("player2", "W2");
    board.setOwner("player3", "S3");
    expect(() => board.setOwner("player4", "N4")).toThrow(
      "No player named player4 is playing"
    );
    expect(() => board.setOwner("player1", "Q3")).toThrow(
      "No property named Q3 exists"
    );
  });
  it("will display the correct info to the correct owners", () => {
    const boardState = board.getBoardState();
    const cherryPicker = (
      obj: Record<string, Record<string, any>>,
      firstSetProps: string[],
      secondSetProps: string[]
    ) => {
      return mapValues(pick(obj, firstSetProps), (value) =>
        pick(value, secondSetProps)
      );
    };
    const anonymous = cherryPicker(
      board.getBoardState(),
      ["E1", "W2", "S3", "N4"],
      ["name", "cardDisplay", "owner"]
    );
    console.log(Object.keys(anonymous))

    const p1 = cherryPicker(
      board.getBoardState("player1"),
      ["E1", "W2", "S3", "N4"],
      ["name", "cardDisplay", "owner"]
    );
    const p2 = cherryPicker(
      board.getBoardState("player2"),
      ["E1", "W2", "S3", "N4"],
      ["name", "cardDisplay", "owner"]
    );
    const p3 = cherryPicker(
      board.getBoardState("player3"),
      ["E1", "W2", "S3", "N4"],
      ["name", "cardDisplay", "owner"]
    );
    ["E1", "W2", "S3", "N4"].forEach((property) => {
      if (property !== "E1") {
        expect(p1[property]).toEqual(anonymous[property]);
      }
      if (property !== "W2") {
        expect(p2[property]).toEqual(anonymous[property]);
      }
      if (property !== "S3") {
        expect(p3[property]).toEqual(anonymous[property]);
      }
    });
    ["E1", "W2", "S3", "N4"].forEach((property) => {
      expect(anonymous[property].cardDisplay).toBe("??")
    })
    expect(anonymous["E1"].cardDisplay).toBe("??");
    expect(p1.E1.cardDisplay).toBe("5♠");
    expect(p2.W2.cardDisplay).toBe("2♦");
    expect(p3.S3.cardDisplay).toBe("8♥");
    expect(anonymous).toEqual({
      E1: {
        cardDisplay: "??",
        name: "E1",
        owner: "player1",
      },
      N4: {
        cardDisplay: "??",
        name: "N4",
        owner: null,
      },
      S3: {
        cardDisplay: "??",
        name: "S3",
        owner: "player3",
      },
      W2: {
        cardDisplay: "??",
        name: "W2",
        owner: "player2",
      },
    });
  });
});
