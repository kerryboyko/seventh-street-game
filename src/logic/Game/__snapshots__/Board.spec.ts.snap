// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Board Board.getBoardState() can get a board state 1`] = `
{
  "E1": {
    "cardDisplay": "??",
    "name": "E1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E2": {
    "cardDisplay": "??",
    "name": "E2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E3": {
    "cardDisplay": "??",
    "name": "E3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E4": {
    "cardDisplay": "??",
    "name": "E4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E5": {
    "cardDisplay": "??",
    "name": "E5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E6": {
    "cardDisplay": "??",
    "name": "E6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "E7": {
    "cardDisplay": "??",
    "name": "E7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N1": {
    "cardDisplay": "??",
    "name": "N1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N2": {
    "cardDisplay": "??",
    "name": "N2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N3": {
    "cardDisplay": "??",
    "name": "N3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N4": {
    "cardDisplay": "??",
    "name": "N4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N5": {
    "cardDisplay": "??",
    "name": "N5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N6": {
    "cardDisplay": "??",
    "name": "N6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "N7": {
    "cardDisplay": "??",
    "name": "N7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S1": {
    "cardDisplay": "??",
    "name": "S1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S2": {
    "cardDisplay": "??",
    "name": "S2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S3": {
    "cardDisplay": "??",
    "name": "S3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S4": {
    "cardDisplay": "??",
    "name": "S4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S5": {
    "cardDisplay": "??",
    "name": "S5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S6": {
    "cardDisplay": "??",
    "name": "S6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "S7": {
    "cardDisplay": "??",
    "name": "S7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W1": {
    "cardDisplay": "??",
    "name": "W1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W2": {
    "cardDisplay": "??",
    "name": "W2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W3": {
    "cardDisplay": "??",
    "name": "W3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W4": {
    "cardDisplay": "??",
    "name": "W4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W5": {
    "cardDisplay": "??",
    "name": "W5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W6": {
    "cardDisplay": "??",
    "name": "W6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
  "W7": {
    "cardDisplay": "??",
    "name": "W7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
    "type": "PROPERTY",
  },
}
`;

exports[`class Board constructor() generates a neighborhood 1`] = `
{
  "deck": Deck {
    "cards": [
      Card {
        "rank": 10,
        "suit": "c",
      },
      Card {
        "rank": 11,
        "suit": "d",
      },
      Card {
        "rank": 7,
        "suit": "c",
      },
      Card {
        "rank": 9,
        "suit": "h",
      },
      Card {
        "rank": 3,
        "suit": "s",
      },
      Card {
        "rank": 6,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "s",
      },
      Card {
        "rank": 5,
        "suit": "s",
      },
      Card {
        "rank": 13,
        "suit": "d",
      },
      Card {
        "rank": 9,
        "suit": "d",
      },
      Card {
        "rank": 5,
        "suit": "c",
      },
      Card {
        "rank": 13,
        "suit": "h",
      },
      Card {
        "rank": 7,
        "suit": "d",
      },
      Card {
        "rank": 4,
        "suit": "d",
      },
      Card {
        "rank": 10,
        "suit": "h",
      },
      Card {
        "rank": 14,
        "suit": "h",
      },
      Card {
        "rank": 8,
        "suit": "h",
      },
      Card {
        "rank": 6,
        "suit": "d",
      },
      Card {
        "rank": 11,
        "suit": "s",
      },
      Card {
        "rank": 13,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "c",
      },
      Card {
        "rank": 4,
        "suit": "c",
      },
      Card {
        "rank": 2,
        "suit": "d",
      },
      Card {
        "rank": 3,
        "suit": "c",
      },
      Card {
        "rank": 13,
        "suit": "c",
      },
      Card {
        "rank": 8,
        "suit": "d",
      },
      Card {
        "rank": 2,
        "suit": "h",
      },
      Card {
        "rank": 10,
        "suit": "d",
      },
      Card {
        "rank": 12,
        "suit": "h",
      },
      Card {
        "rank": 8,
        "suit": "s",
      },
      Card {
        "rank": 12,
        "suit": "c",
      },
      Card {
        "rank": 4,
        "suit": "s",
      },
      Card {
        "rank": 3,
        "suit": "h",
      },
      Card {
        "rank": 12,
        "suit": "d",
      },
      Card {
        "rank": 3,
        "suit": "d",
      },
      Card {
        "rank": 2,
        "suit": "c",
      },
      Card {
        "rank": 5,
        "suit": "h",
      },
      Card {
        "rank": 9,
        "suit": "c",
      },
      Card {
        "rank": 8,
        "suit": "c",
      },
      Card {
        "rank": 6,
        "suit": "c",
      },
      Card {
        "rank": 10,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "d",
      },
      Card {
        "rank": 7,
        "suit": "s",
      },
      Card {
        "rank": 2,
        "suit": "s",
      },
      Card {
        "rank": 12,
        "suit": "s",
      },
      Card {
        "rank": 11,
        "suit": "h",
      },
      Card {
        "rank": 11,
        "suit": "c",
      },
      Card {
        "rank": 5,
        "suit": "d",
      },
      Card {
        "rank": 6,
        "suit": "h",
      },
      Card {
        "rank": 4,
        "suit": "h",
      },
      Card {
        "rank": 9,
        "suit": "s",
      },
      Card {
        "rank": 7,
        "suit": "h",
      },
    ],
    "cursor": 28,
    "dealCard": [Function],
    "dealCards": [Function],
    "length": 52,
    "rng": [Function],
    "seed": "seventhst test",
  },
  "neighborhood": {
    "E0": Property {
      "card": null,
      "getState": [Function],
      "name": "E0",
      "owner": null,
      "price": 0,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "E1": Property {
      "card": Card {
        "rank": 5,
        "suit": "s",
      },
      "getState": [Function],
      "name": "E1",
      "owner": null,
      "price": 100,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E2": Property {
      "card": Card {
        "rank": 13,
        "suit": "d",
      },
      "getState": [Function],
      "name": "E2",
      "owner": null,
      "price": 200,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E3": Property {
      "card": Card {
        "rank": 9,
        "suit": "d",
      },
      "getState": [Function],
      "name": "E3",
      "owner": null,
      "price": 300,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E4": Property {
      "card": Card {
        "rank": 5,
        "suit": "c",
      },
      "getState": [Function],
      "name": "E4",
      "owner": null,
      "price": 400,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E5": Property {
      "card": Card {
        "rank": 13,
        "suit": "h",
      },
      "getState": [Function],
      "name": "E5",
      "owner": null,
      "price": 500,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E6": Property {
      "card": Card {
        "rank": 7,
        "suit": "d",
      },
      "getState": [Function],
      "name": "E6",
      "owner": null,
      "price": 600,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E7": Property {
      "card": Card {
        "rank": 4,
        "suit": "d",
      },
      "getState": [Function],
      "name": "E7",
      "owner": null,
      "price": 700,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N0": Property {
      "card": null,
      "getState": [Function],
      "name": "N0",
      "owner": null,
      "price": 0,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "N1": Property {
      "card": Card {
        "rank": 10,
        "suit": "c",
      },
      "getState": [Function],
      "name": "N1",
      "owner": null,
      "price": 100,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N2": Property {
      "card": Card {
        "rank": 11,
        "suit": "d",
      },
      "getState": [Function],
      "name": "N2",
      "owner": null,
      "price": 200,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N3": Property {
      "card": Card {
        "rank": 7,
        "suit": "c",
      },
      "getState": [Function],
      "name": "N3",
      "owner": null,
      "price": 300,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N4": Property {
      "card": Card {
        "rank": 9,
        "suit": "h",
      },
      "getState": [Function],
      "name": "N4",
      "owner": null,
      "price": 400,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N5": Property {
      "card": Card {
        "rank": 3,
        "suit": "s",
      },
      "getState": [Function],
      "name": "N5",
      "owner": null,
      "price": 500,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N6": Property {
      "card": Card {
        "rank": 6,
        "suit": "s",
      },
      "getState": [Function],
      "name": "N6",
      "owner": null,
      "price": 600,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N7": Property {
      "card": Card {
        "rank": 14,
        "suit": "s",
      },
      "getState": [Function],
      "name": "N7",
      "owner": null,
      "price": 700,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S0": Property {
      "card": null,
      "getState": [Function],
      "name": "S0",
      "owner": null,
      "price": 0,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "S1": Property {
      "card": Card {
        "rank": 10,
        "suit": "h",
      },
      "getState": [Function],
      "name": "S1",
      "owner": null,
      "price": 100,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S2": Property {
      "card": Card {
        "rank": 14,
        "suit": "h",
      },
      "getState": [Function],
      "name": "S2",
      "owner": null,
      "price": 200,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S3": Property {
      "card": Card {
        "rank": 8,
        "suit": "h",
      },
      "getState": [Function],
      "name": "S3",
      "owner": null,
      "price": 300,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S4": Property {
      "card": Card {
        "rank": 6,
        "suit": "d",
      },
      "getState": [Function],
      "name": "S4",
      "owner": null,
      "price": 400,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S5": Property {
      "card": Card {
        "rank": 11,
        "suit": "s",
      },
      "getState": [Function],
      "name": "S5",
      "owner": null,
      "price": 500,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S6": Property {
      "card": Card {
        "rank": 13,
        "suit": "s",
      },
      "getState": [Function],
      "name": "S6",
      "owner": null,
      "price": 600,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S7": Property {
      "card": Card {
        "rank": 14,
        "suit": "c",
      },
      "getState": [Function],
      "name": "S7",
      "owner": null,
      "price": 700,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W0": Property {
      "card": null,
      "getState": [Function],
      "name": "W0",
      "owner": null,
      "price": 0,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "W1": Property {
      "card": Card {
        "rank": 4,
        "suit": "c",
      },
      "getState": [Function],
      "name": "W1",
      "owner": null,
      "price": 100,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W2": Property {
      "card": Card {
        "rank": 2,
        "suit": "d",
      },
      "getState": [Function],
      "name": "W2",
      "owner": null,
      "price": 200,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W3": Property {
      "card": Card {
        "rank": 3,
        "suit": "c",
      },
      "getState": [Function],
      "name": "W3",
      "owner": null,
      "price": 300,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W4": Property {
      "card": Card {
        "rank": 13,
        "suit": "c",
      },
      "getState": [Function],
      "name": "W4",
      "owner": null,
      "price": 400,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W5": Property {
      "card": Card {
        "rank": 8,
        "suit": "d",
      },
      "getState": [Function],
      "name": "W5",
      "owner": null,
      "price": 500,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W6": Property {
      "card": Card {
        "rank": 2,
        "suit": "h",
      },
      "getState": [Function],
      "name": "W6",
      "owner": null,
      "price": 600,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W7": Property {
      "card": Card {
        "rank": 10,
        "suit": "d",
      },
      "getState": [Function],
      "name": "W7",
      "owner": null,
      "price": 700,
      "revealToPlayer": [Function],
      "revealed": false,
      "scored": false,
      "setOwner": [Function],
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
  },
}
`;
